src/sched/job_runner.cc

```diff
@@ -196,6 +199,32 @@ void JobRunnerStrand::PushToRunnerIfNeeded(const bool is_in_running_job) {
     }
 }
 
+// Try running job immediately in the same thread.
+//
+// To ensure correctness without performance regression,
+// We only do a quick check to see if it can be run.
+// If not sure, fall back to the normal AddJob.
+JobRunner::TryRunImmediately::State JobRunnerStrand::AddJob(
+    std::function<void(JobAliveTokenPtr&&)>&& job,
+    JobRunner::TryRunImmediately) {
+    // Decision(D/d). It is protected by lock and it decides whether Push is needed.
+    //
+    // Proceeds only if D+ (changes the `has_ready_job_` value from false to true), and an empty pending queue.
+    // It will not break the order because the pending queue is empty at the point we check.
+    {
+        std::unique_lock lck(hybrid_spin_mtx_, std::try_to_lock);
+        if (!lck.owns_lock() || has_ready_job_ || !pending_jobs_.empty()) {
+            lck.unlock();
+            return AddJob(std::move(job)) ? TryRunImmediately::State::ENQUEUED : TryRunImmediately::State::FAILED;
```

Comment:
(Don't have to be in this PR)

We can be smarter about this, e.g. keep a timed flag so that if a strand/queue is used for async recently, later job will still be async regardlessly. This may have perf benefit on misclassified scheduling.

Reply:
Sounds like an interesting topic, definitely be a follow-up. Maybe assign an issue to me, please. You may also dump your thought into it.